src/parser/mod.rs

```diff
@@ -9225,8 +9225,63 @@ impl<'a> Parser<'a> {
     }
 
     pub fn parse_use(&mut self) -> Result<Statement, ParserError> {
-        let db_name = self.parse_identifier(false)?;
-        Ok(Statement::Use { db_name })
+        // What should be treated as keyword in given dialect
+        let allowed_keywords = if dialect_of!(self is HiveDialect) {
+            vec![Keyword::DEFAULT]
+        } else if dialect_of!(self is DatabricksDialect) {
+            vec![Keyword::CATALOG, Keyword::DATABASE, Keyword::SCHEMA]
+        } else if dialect_of!(self is SnowflakeDialect) {
+            vec![Keyword::DATABASE, Keyword::SCHEMA]
+        } else {
+            vec![]
+        };
+        let parsed_keyword = self.parse_one_of_keywords(&allowed_keywords);
```

Comment:
Representation wise would it make sense to do something like this instead?

```rust
enum Use {
  Catalog(ObjectName), // e.g. `USE CATALOG foo.bar`
  Schema(ObjectName), // e.g. `USE SCHEMA foo.bar`
  Database(ObjectName), // e.g. `USE DATABASE foo.bar`
  Default, // e.g. `USE DEFAULT`
  Object(ObjectName), // e.g. `USE foo.bar`
}
Statement::Use(Use)
```

I'm thinking that could be clearer what's expected, and likely simplify the impl since we would be able to avoid  the manual parsing of object names

Reply:
Do i correctly understand what you propose: you think we could modify the `src/ast/mod.rs:Statement.Use` to accept a single argument, this new `Use` Enum with different values for each syntax used? 

Initially, I considered using an enum instead of a plain String for the keyword, or even taking it a step further with an enum for each syntax, like you propose. However, I might be mistaken, but from the user’s perspective, when later implementing a Visitor for this Use statement, all I care about is easy access to the current database and schema set and the specific syntax used doesn’t really matter. Correct?

This is also my first time writing anything in Rust, so if there are best practices that support your solution, please let me know. I'm not very familiar with them yet, and I'd be happy to make adjustments.

Reply:
Ah yeah the proposal is indeed to modify `Statement::Use` and so that it accepts the enum. Exactly, the dialect/syntax used won't matter after the result comes out of the parser so the visitor would only look at the enum variant and handle each variant accordingly

Reply:
I think this kind of switch from `Statement::Use(db_name)` to `Statement::Use(UseEnum)` would also be breaking change for those who already implemented their Visitors right? No sure what's the deprecation policy, but this is something to consider as well when doing such change. Given everything we've discussed, do you think this would be still a good approach to add that Enum? I'm open to your thoughts.

Reply:
Yeah I think we can still add it. It would be a breaking change, given the stage of the project we do make those when needed unfortunately. Ideally/usually they're mechanical changes to adopt but we try to keep them small in scope in any case to make it easier for downstream crates to adapt to the changes that go into each release.
Specifically for the `Statement` enum there's ongoing work [here](https://github.com/sqlparser-rs/sqlparser-rs/issues/1204) to move away from unnamed structs as variants so that this would likely be an eventual change to the `Statement::Use` variant

Reply:
I've started implementing it and I have one more question, and two examples to show what i mean. Hopefully it's clear:

First, using [Databricks](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-use-schema.html) example, users can provide query like:
`USE foo;` -> It sets SCHEMA to `foo`, but most of dialects would set DATABASE to `foo` with that query.
`USE SCHEMA foo;` -> It also sets SCHEMA to `foo`

After we switch to the Enum we would have:
`USE foo;` -> `UseEnum(Object(foo))`
`USE SCHEMA foo;` -> `UseEnum(Schema(foo))`

When implementing Visitor, the User would have to know that Object(foo) in Databricks sets the SCHEMA and f.e. in MySQL sets the DATABASE.

Then, using [Snowflake](https://docs.snowflake.com/en/sql-reference/sql/use-schema) example, users can provide
`USE SCHEMA foo.bar;` -> It will set DATABASE to `foo` and SCHEMA to `bar`, but f.e. Databricks allow a single identifier after SCHEMA keyword (e.g. only `foo` not `foo.bar`). 

After we switch to the Enum you proposed, we would have:
`USE SCHEMA foo.bar;` -> `UseEnum(Schema(foo.bar))`

When implementing Visitor, the User can't assume that Schema contains a single identifier (of course w still type it as (Schema(ObjectName)) so there is no confusion), but he has to do more parsing of Schema(foo.bar) to still account for different possibilities depending on the Dialect.

We could also possibly translate this query to Object() and not Schema() `USE SCHEMA foo.bar;` -> `UseEnum(Object(foo.bar))`, to show that we are dealing with full identifier, but the we lose the information about the keyword being used.

To sum up, i'm not saying it's wrong, just wanted to make sure that this is the desired division of responsibility - that the result returned by the parser still requires parsing effort from the user and knowing what is allowed / possible in different dialects. Or maybe I'm missing something here? Maybe my assumption writing current approach was wrong, that i wanted to deliver parsed information already and keep all the dialect-specific parsing inside 😄

Reply:
Regardless of my doubts I implemented the desired approach with Enum. Let me know if it looks better now @iffyio

Reply:
The changes look good to me thanks! Re the databricks example, yeah I think that would be expected behavior - goal of the parser is to try to faithfully represent the SQL syntax - semantics is deferred to the consumer of the output AST. So that if the input SQL is `USE foo;`, then the expected parser output indeed becomes `Use(Object(foo))`, an output `Use(Schema(foo))` I think could be surprising parser behavior.

> After we switch to the Enum you proposed, we would have:
> `USE SCHEMA foo.bar; -> UseEnum(Schema(foo.bar))`

Similarly, that looks like expected parser output to me.

Not exactly the same scenario but potentially related section from the project's goals re syntax vs semantics: https://github.com/sqlparser-rs/sqlparser-rs?tab=readme-ov-file#syntax-vs-semantics